plugins/woocommerce/src/Internal/ProductFilters/QueryClauses.php

```diff
@@ -0,0 +1,401 @@
+<?php
+/**
+ * QueryClauses class file.
+ */
+
+declare(strict_types=1);
+
+namespace Automattic\WooCommerce\Internal\ProductFilters;
+
+use Automattic\WooCommerce\Internal\ProductAttributesLookup\LookupDataStore;
+use Automattic\WooCommerce\Internal\ProductFilters\Interfaces\QueryClausesGenerator;
+use WC_Tax;
+use WC_Cache_Helper;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class for filter clauses.
+ */
+class QueryClauses implements QueryClausesGenerator {
+
+	/**
+	 * Add conditional query clauses based on the filter params in query vars.
+	 *
+	 * @param array     $args     Query args.
+	 * @param \WP_Query $wp_query WP_Query object.
+	 * @return array
+	 */
+	public function add_query_clauses( $args, $wp_query ) {
+		// Rating filter is handled by tax query.
+		if ( $wp_query->get( 'filter_stock_status' ) ) {
+			$stock_statuses = trim( $wp_query->get( 'filter_stock_status' ) );
+			$stock_statuses = explode( ',', $stock_statuses );
+			$stock_statuses = array_filter( $stock_statuses );
+
+			$args = $this->add_stock_clauses( $args, $stock_statuses );
+		}
+
+		if ( $wp_query->get( 'min_price' ) || $wp_query->get( 'max_price' ) ) {
+			$price_range = array(
+				'min_price' => $wp_query->get( 'min_price' ),
+				'max_price' => $wp_query->get( 'max_price' ),
+			);
+			$price_range = array_filter( $price_range );
+			$args        = $this->add_price_clauses( $args, $price_range );
+		}
+
+		$args = $this->add_attribute_clauses(
+			$args,
+			$this->get_chosen_attributes( $wp_query->query_vars )
+		);
+
+		return $args;
+	}
+
+	/**
+	 * Add query clauses for stock filter.
+	 *
+	 * @param array $args           Query args.
+	 * @param array $stock_statuses Stock statuses to be queried.
+	 * @return array
+	 */
+	public function add_stock_clauses( $args, $stock_statuses ) {
+		if ( empty( $stock_statuses ) ) {
+			return $args;
+		}
+
+		$stock_statuses = array_intersect(
+			array_map( 'esc_sql', $stock_statuses ),
+			array_keys( wc_get_product_stock_status_options() )
+		);
+
+		$args['join']   = $this->append_product_sorting_table_join( $args['join'] );
+		$args['where'] .= ' AND wc_product_meta_lookup.stock_status IN ("' . implode( '","', $stock_statuses ) . '")';
```

Comment:
I thought `esc_sql( implode( '","', $stock_statuses ) )` would work? Even though the [docs](https://developer.wordpress.org/reference/functions/esc_sql/) say "It will only escape values to be used in strings in the query", that's exactly what's happening here, no? We only need to escape `$stock_statuses`, which will be between quotes. :thinking:

Reply:
We need to pass multiple values (string) to `IN`, not a comma-separate string, which is only one value.

Reply:
I'm not sure to understand why `esc_sql( implode( '","', $stock_statuses ) )` wouldn't work. :sweat_smile: But it was a small nit, so I'm happy to leave it as-is. Definitely not a blocker.